react-components/src/components/Reveal3DResources/Image360StylingHandler.ts

```diff
@@ -0,0 +1,136 @@
+
+type CollectionStyleGroupWithAssetSet = {
+  styling: Image360AnnotationAppearance;
+  assetIdSet: Set<number>;
+};
+
+export class Image360StylingHandler {
+  private readonly _collections: Image360AddOptionsWithModel[];
+  private _defaultStyling: DefaultResourceStyling['image360'] = undefined;
+  private _instanceStyling: Image360AssetStylingGroup[] = [];
+  private readonly _viewer: Cognite3DViewer;
+
+  private readonly _previousImage360StylingCallbacks = new Map<
+    string,
+    (image: Image360, imageRevision: Image360Revision) => void
+  >();
+
+  constructor(viewer: Cognite3DViewer) {
+    this._viewer = viewer;
+    this._collections = [];
+  }
+
+  public setCommonStyling(
+    defaultStyling: DefaultResourceStyling['image360'],
+    instanceStyling: InstanceStylingGroup[]
+  ): void {
+    this._defaultStyling = defaultStyling;
+    this._instanceStyling = instanceStyling.filter(isImage360AssetStylingGroup);
+  }
+
+  async setCollections(models: AddOptionsWithModel[]): Promise<void> {
+    const collections = models.filter((model): model is Image360AddOptionsWithModel =>
+      is360ImageCollectionOptions(model.addOptions)
+    );
+
+    this._collections.push(...collections);
+  }
+
+  public async update360ImageStylingCallback(collection: Image360Collection): Promise<void> {
+    const collectionId = collection.id;
+    const prevCallback = this._previousImage360StylingCallbacks.get(collectionId);
+    if (prevCallback !== undefined) {
+      collection.off('image360Entered', prevCallback);
+    }
+
+    const styleGroups = this.compute360ImageStyleGroups();
+    const collectionStyleGroupsWithSet = styleGroups.map((group) => ({
+      styling: group.style,
+      assetIdSet: new Set(group.assetIds)
+    }));
+
+    const defaultStyling = this._defaultStyling?.default;
+
+    const onEnteredStylingCallback: Image360EnteredDelegate = async (_image, imageRevision) => {
+      const annotations = await imageRevision.getAnnotations();
+      annotations.forEach((annotation) => {
+        styleAnnotation(annotation, defaultStyling, collectionStyleGroupsWithSet);
+      });
+    };
+
+    collection.on('image360Entered', onEnteredStylingCallback);
+    this._previousImage360StylingCallbacks.set(collectionId, onEnteredStylingCallback);
+
+    this.styleCurrentlyEntered360Image(onEnteredStylingCallback);
+  }
+
+  private styleCurrentlyEntered360Image(stylingCallback: Image360EnteredDelegate): void {
+    const currentlyActiveImage = this._viewer.getActive360ImageInfo()?.image360;
+
+    if (currentlyActiveImage === undefined) {
+      return;
+    }
+
+    stylingCallback(currentlyActiveImage, currentlyActiveImage.getActiveRevision());
+  }
+
+  private compute360ImageStyleGroups(): Image360AnnotationStyleGroup[] {
+    return this._instanceStyling?.map((group) => {
+      return { assetIds: group.assetIds, style: group.style.image360 };
+    });
+  }
+}
+
+function styleAnnotation(
```

Comment:
Any reason to have this function outside the class

Reply:
More of a personal preference - if a function doesn't need (or change) the state of a class, I put it outside. It makes it conceptually easier to reason about, since you know all its inputs. I can move it if you disagree strongly